build,server: porting client interceptors to drpc

shaikzakiriitm opened the pull request:
build,server: porting client interceptors to drpc

Added dependency on drpc version which contains support for drpc interceptors.
Created drpc variant of client interceptors. 

Epic: [CRDB-49359](https://cockroachlabs.atlassian.net/browse/CRDB-49359)
Fixes: https://github.com/cockroachdb/cockroach/issues/144374
Part of: [CRDB-49358](https://cockroachlabs.atlassian.net/browse/CRDB-49358)

shaikzakiriitm committed:
build,server: add dependency of drpc client interceptors

Drpc didn't have support for interceptors. For us to use existing
grpc client interceptors with drpc client, we need to import
the package drpcclient from cockroach's fork of drpc repo.
As part of this change, we are importing external drpcclient
package.
This import brings in definitions of drpc client interceptors
and enables migration of grpc interceptors to drpc ones.

Epic: CRDB-49359
Fixes: #144374
Part of: CRDB-49358

Release note: None

shaikzakiriitm committed:
server: port callCountInterceptor to drpc

callCountInterceptor in grpc_gateway is used to increment telemetry counters
for each grpc method.
Here we are porting this interceptor to drpc.
Created a drpc interceptor for callCount, but its switchable with grpc
counterpart depending upon feature flag to enable drpc.

Epic: CRDB-49358
Fixes: #144374

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147592)
<!-- Reviewable:end -->

shaikzakiriitm committed:
rpc: port tracing client interceptor to drpc

To enable tracing on rpc calls, a client side unary and stream interceptor
is implemented adhering to grpc interfaces.
As part of this change, client-side tracing interceptors are
implemented adhering to drpc interfaces.
This is needed to enable tracing in rpc calls based on drpc.

Epic: CRDB-49359
Fixes: #144372
Part of: CRDB-49356

Release note: None

shaikzakiriitm committed:
fixup! rpc: port tracing client interceptor to drpc

shaikzakiriitm commented:
Closing this pr in favor of other prs where we reordered the commits.
